extra/mariadbfrm/mariadb_frm.cc

```diff
@@ -0,0 +1,531 @@
+/* Copyright (c) 2024, MariaDB
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1335 USA */
+
+/**
+  @file extra/mariadb_frm.cc
+  FRM file parser utility - extracts table structure from .frm files
+*/
+
+#include "mariadb.h"
+
+#include "mariadb.h"
+#include "my_sys.h"
+#include "mysql_com.h"
+#include "m_ctype.h"
+#include "field.h"
+#include "sql_class.h"
+#include "sql_error.h"
+
+
+extern mysql_mutex_t LOCK_start_thread, LOCK_status, LOCK_global_system_variables, LOCK_user_conn, LOCK_thread_id;
+extern mysql_cond_t COND_start_thread;
+extern struct system_variables global_system_variables;
+
+PSI_mutex_key key_LOCK_start_thread = 0;
+PSI_mutex_key key_LOCK_status = 0;
+PSI_mutex_key key_LOCK_global_system_variables = 0;
+PSI_mutex_key key_LOCK_user_conn = 0;
+PSI_mutex_key key_LOCK_thread_id = 0;
+PSI_cond_key key_COND_start_thread = 0;
+PSI_mutex_key key_LOCK_plugin = 0;
+
+
+bool plugins_are_initialized = false;
+mysql_mutex_t LOCK_plugin;
+struct st_plugin_int **plugin_array = nullptr;
+uint plugin_array_size = 0;
+
+
+CHARSET_INFO *system_charset_info = nullptr;
+CHARSET_INFO *files_charset_info = nullptr;
+CHARSET_INFO *national_charset_info = nullptr;
+CHARSET_INFO *table_alias_charset = nullptr;
+CHARSET_INFO *character_set_filesystem = nullptr;
+CHARSET_INFO *default_charset_info = nullptr;
+
+extern int mysqld_server_started;
+extern int mysqld_server_initialized;
+extern int sf_leaking_memory;
+
+
+static handlerton mock_hton;
+
+static st_maria_plugin mock_plugin = {
+  .type = MYSQL_STORAGE_ENGINE_PLUGIN,
```

Comment:
Formally, designated initialization is only possible since c++20. I think nothing can stop us to enable it in cmake. Please try to make it, but only for frm_parser!